cpp/include/cudf/detail/aggregation/aggregation.cuh

```diff
@@ -252,59 +242,69 @@ struct update_target_element<Source,
+                             column_device_view source,
                              size_type source_index) const noexcept
   {
-    auto const keys  = d_dictionary.child(cudf::dictionary_column_view::keys_column_index);
-    auto const value = keys.element<KeyType>(
-      static_cast<cudf::size_type>(d_dictionary.element<dictionary32>(source_index)));
-    using Target = target_type_t<KeyType, aggregation::SUM>;
-    atomicAdd(&target.element<Target>(target_index), static_cast<Target>(value));
+    update_target_element<Source, k, target_has_nulls, source_has_nulls>{}(
+      target, target_index, source, source_index);
   }
-  template <typename KeyType,
-            std::enable_if_t<!is_fixed_width<KeyType>() || is_fixed_point<KeyType>()>* = nullptr>
-  __device__ void operator()(mutable_column_device_view& target,
+  template <typename Source,
+            aggregation::Kind k,
+            std::enable_if_t<is_dictionary<Source>()>* = nullptr>
+  __device__ void operator()(mutable_column_device_view target,
                              size_type target_index,
-                             column_device_view& d_dictionary,
-                             size_type source_index) const noexcept {};
+                             column_device_view source,
+                             size_type source_index) const noexcept
+  {
+  }
 };
 
 /**
- * @brief Specialization function for dictionary type and aggregation SUM.
+ * @brief Specialization function for dictionary type and aggregations.
  *
  * @tparam target_has_nulls Indicates presence of null elements in `target`
  * @tparam source_has_nulls Indicates presence of null elements in `source`.
  */
-template <bool target_has_nulls, bool source_has_nulls>
-struct update_target_element<dictionary32, aggregation::SUM, target_has_nulls, source_has_nulls> {
+template <aggregation::Kind k, bool target_has_nulls, bool source_has_nulls>
+struct update_target_element<
+  dictionary32,
+  k,
+  target_has_nulls,
+  source_has_nulls,
+  std::enable_if_t<not(k == aggregation::ARGMIN or k == aggregation::ARGMAX or
+                       k == aggregation::COUNT_VALID or k == aggregation::COUNT_ALL)>> {
   __device__ void operator()(mutable_column_device_view target,
                              size_type target_index,
                              column_device_view source,
                              size_type source_index) const noexcept
   {
     if (source_has_nulls and source.is_null(source_index)) { return; }
 
-    type_dispatcher(source.child(cudf::dictionary_column_view::keys_column_index).type(),
-                    update_target_from_dictionary_sum{},
-                    target,
-                    target_index,
-                    source,
-                    source_index);
+    dispatch_type_and_aggregation(
+      source.child(cudf::dictionary_column_view::keys_column_index).type(),
+      k,
+      update_target_from_dictionary<target_has_nulls, source_has_nulls>{},
```

Comment:
```suggestion
      update_target_from_dictionary<target_has_nulls, false>{},
```
The dictionary keys never contain nulls so this template parameter could also be removed from `update_target_from_dictionary`. This may generate slightly less code.